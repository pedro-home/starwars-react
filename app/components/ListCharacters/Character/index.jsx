import React from 'react';
import { Panel, List } from 'rsuite';
import Info from './Info';

const Component = ({ data, imageUrl }) => {
  return (
    <Panel shaded bordered bodyFill>
      <img src={imageUrl} height="240" />
      <Panel>
        <List>
          {Object.entries(data).map((key, value) => (
            <List.Item key={key}>
              <Info type={key} value={value} />
            </List.Item>
          ))}
        </List>
      </Panel>
    </Panel>
  );
};

export default Component;
