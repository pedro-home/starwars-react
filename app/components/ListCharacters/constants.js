export const OFFSET_SCROLL = 200;
