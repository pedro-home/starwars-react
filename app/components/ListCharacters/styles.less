.list-characters {
  margin: -1.25rem;

  > * {
    padding: 1.25rem;
  }
}

@import './Character/styles';
