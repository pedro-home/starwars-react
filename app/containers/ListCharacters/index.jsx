import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { getErrorMessage, getData, isWaiting } from 'providers/StarWars/selectors';
import ListCharacters from 'components/ListCharacters';

const Container = ({ error, data, isLoading }) => {
  return <ListCharacters error={error} data={data} isLoading={isLoading} />;
};

const mapStateToProps = createStructuredSelector({
  error: getErrorMessage(),
  isLoading: isWaiting(),
  data: getData(),
});

export default connect(mapStateToProps)(Container);
