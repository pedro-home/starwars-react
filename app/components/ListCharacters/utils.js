import femaleLogo from './assets/female.jpg';
import maleLogo from './assets/male.jpg';
import hermaphroditeLogo from './assets/hermaphrodite.jpg';
import droidLogo from './assets/droid.jpg';
import { OFFSET_SCROLL } from './constants';

export const getCharacterImage = name => {
  let logo;
  switch (name) {
    case 'female':
      logo = femaleLogo;
      break;
    case 'male':
      logo = maleLogo;
      break;
    case 'hermaphrodite':
      logo = hermaphroditeLogo;
      break;
    default:
      logo = droidLogo;
  }

  return logo;
};

export const hasEndedScroll = (offset = OFFSET_SCROLL) => {
  return window.scrollY + window.innerHeight >= document.body.scrollHeight - offset;
};
