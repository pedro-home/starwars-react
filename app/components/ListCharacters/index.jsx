import React, { useEffect } from 'react';
import { FlexboxGrid, Loader, Col } from 'rsuite';
import PropTypes from 'prop-types';
import Quote from 'components/Quote';
import ContainerCenter from 'components/ContainerCenter';
import Character from './Character';
import { getCharacterImage, hasEndedScroll } from './utils';

const Component = ({ error, data, isLoading, nextFetch, fetch }) => {
  useEffect(() => {
    if (!nextFetch) {
      return;
    }

    const handleScroll = () => {
      if (hasEndedScroll() && !isLoading) {
        fetch(nextFetch);
      }
    };

    window.addEventListener('scroll', handleScroll);

    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, [nextFetch, fetch, isLoading]);

  let component = (
    <>
      <FlexboxGrid className="list-characters">
        {data.map(value => (
          <FlexboxGrid.Item componentClass={Col} md={6} sm={12} colspan={24} key={value.name}>
            <Character data={value} imageUrl={getCharacterImage(value.gender)} />
          </FlexboxGrid.Item>
        ))}
        {isLoading && (
          <FlexboxGrid.Item key="loader" componentClass={Col} colspan={24}>
            <Loader center content="loading..." />
          </FlexboxGrid.Item>
        )}
      </FlexboxGrid>
    </>
  );

  if (error !== '' || (data.length === 0 && !isLoading)) {
    component = (
      <ContainerCenter>
        <Quote text="Search your feelings, you know you to be true" />
      </ContainerCenter>
    );
  }

  return component;
};

Component.defaultProps = {
  data: [],
  isLoading: false,
  fetch: () => {},
  error: '',
};

Component.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  nextFetch: PropTypes.object,
  fetch: PropTypes.func,
};

export default Component;
