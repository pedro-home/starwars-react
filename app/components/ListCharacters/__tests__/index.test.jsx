import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import ListCharacters from '../index';

describe('Given <ListCharacters />', () => {
  let component;
  let onFetch;
  let scrollWindow;

  beforeAll(() => {
    scrollWindow = () => {
      window.scrollTo({ top: document.body.scrollHeight });
      fireEvent.scroll(window);
    };
  });

  describe('When initializes', () => {
    beforeEach(() => {
      const { container } = render(<ListCharacters />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('And should render a center container', () => {
      expect(component.querySelector('.center-container')).toBeTruthy();
    });

    it('And should render a quote', () => {
      expect(component.querySelector('.quote')).toBeTruthy();
    });
  });

  describe('When gets data', () => {
    let data;
    beforeEach(() => {
      data = [{ name: 'Test' }, { name: 'Test2' }];
      const { container } = render(<ListCharacters data={data} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('And should render characters', () => {
      const characters = Array.from(component.querySelectorAll('.character'));
      expect(characters.length).toEqual(data.length);
      expect(characters.map(elem => ({ name: elem.querySelector('.title').innerHTML }))).toStrictEqual(data);
    });
  });

  describe('When fetches data', () => {
    beforeEach(() => {
      onFetch = jest.fn();
    });

    describe('And without next page', () => {
      beforeEach(() => {
        const { container } = render(<ListCharacters fetch={onFetch} />);
        component = container;
        scrollWindow();
      });

      it('Then fetch should not be called', () => {
        expect(onFetch).toHaveBeenCalledTimes(0);
      });
    });

    describe('And during loading', () => {
      beforeEach(() => {
        const { container } = render(<ListCharacters fetch={onFetch} isLoading={true} />);
        component = container;
        scrollWindow();
      });

      it('Then fetch should not be called', () => {
        expect(onFetch).toHaveBeenCalledTimes(0);
      });
    });

    describe('And with next and no loading', () => {
      beforeEach(() => {
        const { container } = render(<ListCharacters fetch={onFetch} isLoading={false} nextFetch={{ page: 2 }} />);
        component = container;
        scrollWindow();
      });

      it('Then fetch should be called', () => {
        expect(onFetch).toHaveBeenCalledTimes(1);
      });
    });

    afterEach(() => {
      window.scrollTo({ top: 0 });
    });
  });

  describe('When is loading', () => {
    beforeEach(() => {
      const { container } = render(<ListCharacters isLoading={true} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('And should render loader', () => {
      expect(component.querySelector('.rs-loader')).toBeTruthy();
    });
  });

  describe('When has an error', () => {
    beforeEach(() => {
      const { container } = render(<ListCharacters error="Ups... This is an error!" data={[{ name: 'Hello' }]} />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });

    it('And should render a quote', () => {
      expect(component.querySelector('.quote')).toBeTruthy();
    });

    it('And should render no characters', () => {
      expect(component.querySelectorAll('.character').length).toEqual(0);
    });
  });
});
