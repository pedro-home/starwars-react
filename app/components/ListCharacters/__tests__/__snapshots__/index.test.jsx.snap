// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <ListCharacters /> When gets data Then should match the snapshot 1`] = `
<div>
  <div
    class="rs-flex-box-grid list-characters rs-flex-box-grid-top rs-flex-box-grid-start"
  >
    <div
      class="rs-flex-box-grid-item rs-flex-box-grid-item-24 rs-col rs-col-sm-12 rs-col-md-6"
    >
      <div
        class="character rs-panel rs-panel-default rs-panel-shaded"
      >
        <div
          class="rs-panel-body rs-panel-body-fill"
        >
          <img
            class="character-image"
            src="droid.jpg"
          />
          <div
            class="rs-list"
          >
            <div
              class="rs-list-item character-entry rs-list-item-md"
              manager="[object Object]"
            >
              <div
                class="rs-list-item-content"
              >
                <h6
                  class="title"
                >
                  Test
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="rs-flex-box-grid-item rs-flex-box-grid-item-24 rs-col rs-col-sm-12 rs-col-md-6"
    >
      <div
        class="character rs-panel rs-panel-default rs-panel-shaded"
      >
        <div
          class="rs-panel-body rs-panel-body-fill"
        >
          <img
            class="character-image"
            src="droid.jpg"
          />
          <div
            class="rs-list"
          >
            <div
              class="rs-list-item character-entry rs-list-item-md"
              manager="[object Object]"
            >
              <div
                class="rs-list-item-content"
              >
                <h6
                  class="title"
                >
                  Test2
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Given <ListCharacters /> When has an error Then should match the snapshot 1`] = `
<div>
  <div
    class="container-center"
  >
    <h4
      class="quote"
    >
      <q>
        Search your feelings, you know you to be true
      </q>
    </h4>
  </div>
</div>
`;

exports[`Given <ListCharacters /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="container-center"
  >
    <h4
      class="quote"
    >
      <q>
        Search your feelings, you know you to be true
      </q>
    </h4>
  </div>
</div>
`;

exports[`Given <ListCharacters /> When is loading Then should match the snapshot 1`] = `
<div>
  <div
    class="rs-flex-box-grid list-characters rs-flex-box-grid-top rs-flex-box-grid-start"
  >
    <div
      class="rs-flex-box-grid-item rs-flex-box-grid-item-24 rs-col"
    >
      <div
        class="rs-loader-wrapper rs-loader-speed-normal rs-loader-center rs-loader-has-content"
      >
        <div
          class="rs-loader"
        >
          <span
            class="rs-loader-spin"
          />
          <span
            class="rs-loader-content"
          >
            loading...
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
